torchtitan/experiments/flux/train.py

```diff
@@ -192,43 +213,308 @@ def train_step(self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor):
 
-        save_image(
-            name=f"image_rank{str(torch.distributed.get_rank())}_{self.step}.png",
-            output_dir=os.path.join(
-                self.job_config.job.dump_folder, self.job_config.eval.save_img_folder
-            ),
-            x=image,
-            add_sampling_metadata=True,
-            prompt=prompt,
+            # Avoid division by zero
+            timestep_counts = torch.maximum(
+                timestep_counts, torch.ones_like(timestep_counts)
+            )
+
+            if (
+                parallel_dims.dp_replicate_enabled
+                or parallel_dims.dp_shard_enabled
+                or parallel_dims.cp_enabled
+            ):
+                # Collect loss sums and counts from all devices
+                ft_pg = (
+                    self.ft_manager.replicate_pg if self.ft_manager.enabled else None
+                )
+                # Use the new dist_collect function to gather tensors across devices
+                global_loss_per_timestep = dist_utils.dist_collect(
+                    loss_per_timestep, world_mesh["dp_cp"], ft_pg
+                )
+                global_timestep_counts = dist_utils.dist_collect(
+                    timestep_counts, world_mesh["dp_cp"], ft_pg
+                )
+
+            else:
+                # For single device, just calculate locally
+                global_loss_per_timestep = loss_per_timestep
+                global_timestep_counts = timestep_counts
+
+            # In the future, we could return avg_loss_per_timestep for more detailed reporting
+            return global_loss_per_timestep, global_timestep_counts
+    
+    @record
+    def inference(self, prompts: list[str], bs: int = 1):
+        """
+        Run inference on the Flux model.
+        """
+        self.checkpointer.load(step=self.job_config.checkpoint.load_step)
+        results = []
+        with torch.no_grad():
+            for i in range(0, len(prompts), bs):
+                images = generate_image(
+                    device=self.device,
+                    dtype=self._dtype,
+                    job_config=self.job_config,
+                    model=self.model_parts[0],
+                    prompt=prompts[i : i + bs],
+                    autoencoder=self.autoencoder,
+                    t5_tokenizer=FluxTokenizer(
+                        self.job_config.encoder.t5_encoder,
+                        max_length=self.job_config.encoder.max_t5_encoding_len,
+                    ),
+                    clip_tokenizer=FluxTokenizer(
+                        self.job_config.encoder.clip_encoder, max_length=77
+                    ),
+                    t5_encoder=self.t5_encoder,
+                    clip_encoder=self.clip_encoder,
+                )
+                results.append(images.detach())
+        results = torch.cat(results, dim=0)
+        return results
+
+    def generate_and_save_images(self, inputs) -> torch.Tensor:
```

Comment:
Can we move this function to sampling.py, and reuse some of the functions there?  We could calculate empty_batch in train.py, and pass it to the function call.

In general, we want to make the train.py similar and clean to read